[1단계 - 자동차 경주 구현] 유콩(김유빈) 미션 제출합니다.

kyukong opened the pull request:
안녕하세요! 유콩입니다.
프리코스를 진행하면서 우테코에 가장 기대했던 것이 개인 피드백이었는데, 어느새 첫 리뷰를 받는 날이 왔네요.
교육을 위해 시간을 내주셔서 감사합니다.
피드백 기다리고 있겠습니다!!

***

자동차 경주 구현 미션을 진행하면서 여쭤보고 싶은 것들을 정리해보았습니다.

### 1. 랜덤값에 대한 테스트 진행
자동차 이동 함수로 예를 들 때,

1. 0과 9사이의 값을 랜덤으로 추출함
2. 추출한 값을 기준으로 이동 여부를 반환함
의 과정을 한 메소드에 담아 랜덤 테스트에 대한 어려움이 있었습니다.

제 페어 수달의 아이디어 덕분에 위 메소드를 1번과 2번으로 분리하여 각 랜덤값을 호출하는 테스트 케이스를 추가할 수 있었습니다.

```java
        private boolean hasNext() {
		return generate() > DRIVE_FLAG;
	}

	private int generate() {
		return (int)(Math.random() * RANDOM_MULTIPLE) % RANDOM_DIVISION;
	}
```

```java
	@DisplayName("자동차 랜덤 숫자 범위 확인")
	@Test
	void cars_random_range() {
		//given, when
		int value = generate();
		//then
		assertTrue(value > -1 && value < 11);
	}
```

프리코스 진행 당시 랜덤에 나올 값을 고정하여 추측하고 테스트 케이스를 실행하는 메소드가 있었는데,
혹시 다음과 같은 테스트 메소드를 만든다면 어떤 것을 찾아보는 것이 좋을지 여쭈어보고 싶습니다.

### 2. 상수 처리
지난번 프리코스 때에는 상수를 현재 사용하는 클래스 내에 final 로 설정하여 모아 두었습니다.
그러나 한 클래스 내에 모아두다보니 `상수 개수 증가`, `다른 클래스 내에서의 호출` 등의 문제로 Enum 클래스로 분리하였습니다.

이번 과제에서는 클래스 분리를 잘 나누었는지,
`다른 클래스 내에서의 호출`의 문제가 발생하지 않아 사용하는 클래스 내에 모든 상수를 선언하였습니다.

상수의 개수가 증가함에 따라 해당 클래스 내에 선언하는 것이 좋은지, Enum 클래스로 분리하는 것이 좋은지, 아니면 또 다른 방법이 있는지 알고 싶습니다.

***

아직 많이 부족하여 제대로 질문이 되었는지 모르겠네요..^^;
피드백 기다리고 있겠습니다. 감사합니다!!

her0807 committed:
feat: 계산기 생성

her0807 committed:
test: String에 관한 테스트 추가

her0807 committed:
test: Set에 관한 테스트 추가

her0807 committed:
test: 테스트 케이스를 추가함

her0807 committed:
test: 중간에 공백이나 null 이 들어올 경우

her0807 committed:
test: 값이 문자일 경우 테스트

her0807 committed:
feat: 값이 Null or 빈값일 경우 예외처리

her0807 committed:
feat: 숫자 하나만 들어올 경우 예외 추가

her0807 committed:
feat: , : 구분자 추가

her0807 committed:
feat: 커스텀 구분자가 있을 경우 기능 추가

her0807 committed:
feat: 음수일 경우 예외 추가

her0807 committed:
docs: 기능 구현 목록 추가

kyukong committed:
test: 자동차 이름이 공백일 경우 테스트 추가

kyukong committed:
test: 입력한 자동차 개수가 2개 미만인 경우 테스트 추가

kyukong committed:
test: 자동차 이름이 중복될 경우 테스트 추가

kyukong committed:
test: 자동차 이름의 길이가 5글자를 초과하는 경우 테스트 추가

kyukong committed:
feat: 자동차 이름이 공백인 경우 추가

kyukong committed:
feat: 자동차 개수가 2개 미만인 경우 추가

kyukong committed:
feat: 자동차 이름이 중복될 경우 추가

kyukong committed:
feat: 자동차의 이름이 5글자를 초과하는 경우 추가

kyukong committed:
refactor(Input): 메소드 분리

kyukong committed:
fix: 자동차 입력받는 메소드 수정

- 자동차 이름의 길이 제한 수정(5글자 포함)
- 에러 메시지 출력 추가

kyukong committed:
fix(InputTest): 자동차 입력 테스트 수정

kyukong committed:
refactor: [ERROR] 메시지 추가

kyukong committed:
docs: README 갱신

kyukong committed:
test: 시도횟수 공백 테스트 추가

kyukong committed:
test: 시도횟수 문자 예외 테스트 추가

kyukong committed:
test: 시도횟수 음수 테스트 추가

kyukong committed:
feat: 시도횟수 문자 예외 확인 추가

kyukong committed:
feat: 시도횟수 1미만의 수 확인 추가

kyukong committed:
fix(InputTest): 시도횟수 음수 테스트 시 입력값 타입 수정

- int 타입에서 String 타입으로 수정

kyukong committed:
docs: README 갱신

kyukong committed:
feat: Car 클래스 생성

kyukong committed:
feat: Cars 클래스 생성

kyukong committed:
feat: Attempt 클래스 생성

kyukong committed:
refactor: 데이터 일급객체, 일급컬렉션으로 포장

kyukong committed:
feat: 레이싱 게임을 진행할 RacingCarGame 클래스 생성

kyukong committed:
feat: 자동차 이동 기능 구현

kyukong committed:
feat: 자동차 이동 구현 메소드 실행

kyukong committed:
docs: README 갱신

kyukong committed:
test: 현재 횟수와 시도횟수가 동일한지 확인하는 테스트 추가

kyukong committed:
test: 자동차 위치 이동 테스트 추가

kyukong committed:
feat: 경기 결과 출력

kyukong committed:
docs: README 갱신

kyukong committed:
feat: 최종 우승자 출력

kyukong committed:
refactor: 자동차 위치 이동 테스트 수정

kyukong committed:
fix(Input): 자동차 이름 길이 예외처리 수정

kyukong committed:
refactor: 들여쓰기 수정함에 따라 관련 코드 수정

kyukong committed:
docs: README 갱신

her0807 committed:
refactor(Car): 상수 추출과 메서드 분리

- Car와 Cars에 역할을 명확하게 하기 위해 메서드 이동
- 주요기능인 drive 에 관한 test 수행

her0807 committed:
refactor(Cars): 상수 추출과 자동차 경주, 우승자 추출 로직 테스트

- hasNext() 에 가중된 책임을 generate() 메서드분리로 해결
- 상수 추출
- 랜덤 숫자 범위 test ,우승자 추출 test

her0807 committed:
refactor(RacingCarGame): 조금더 직관적인 네이밍으로 수정, 상수 추출

her0807 committed:
refactor(Input): 상수 추출, 공백과 null error 분리

- 상수선언
- null 과 공백일 경우 에러메세지 상세하게 출력
- 자동차 이름 입력 값에 대한 Null 처리 test 추가
- 반복 횟수 입력 값에 대한 Null 처리 test 추가

KIMSIYOUNG commented on README.md:
```diff
@@ -2,6 +2,54 @@
-- [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+---
+
+> ### 1. 입력
```
대주제를 정해주셔서 어떤 프로그램인지 파악하기 쉽네요 👏

프로그램 자체가 어떤 기능을 하는지 간단한 요약본도 적어주면 더 좋은 프로그램 설명서가 될 것 같아요.

KIMSIYOUNG commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.regex.Pattern;
+
+public class Calculator {
+	public int splitAndSum(String text) {
```
해당 메소드는 2가지 일을 같이 하고 있는 것 같아요🙂

어떻게 분리하면 좋을지 고민해보고, 왜 메소드를 작게 나눠야 하는지에 대해서도 고민해보면 좋을 것 같아요!

KIMSIYOUNG replied:
요 부분에 대해서는 아직 반영이 되지 않은 것 같아요🙂

KIMSIYOUNG commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,75 @@
+		return string.matches("[+-]?\\d*(\\.\\d+)?");
+	}
+
+	private String[] getValues(String text) {
```
해당 메소드의 이름을 맥락없이 보았을 때 어떤 느낌이 드나요? 

`text`를 넣으면 `String 배열`이 나오는데 어떤 스트링배열이 나올지 조금 가늠하기 어려운 것 같아요. 

보다 의도가 드러나는 네이밍을 고민해보면 좋을 것 같아요.

KIMSIYOUNG commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,75 @@
+		}
+	}
+
+	private int add(int[] numbers) {
```
`Add`라는 건 일반적으로 컬렉션에 무언가를 추가한다거나, 기존 특정 값에 어떤것을 더한다로 주로 사용하는 것으로 인지하고 있는데 조금 더 직관적인 네이밍은 없을지 고민해보아요!

KIMSIYOUNG commented on src/main/java/racingcar/domain/Attempt.java:
```diff
@@ -0,0 +1,13 @@
+package racingcar.domain;
+
+public class Attempt {
```
원시값을 포장하여 유의미한 객체를 만드셨네요👏

다만 해당 객체에서 시도 횟수의 경우 음수는 불가능한 것으로 보이는데, 요 부분에 대한 Validation과 테스트도 함께 추가해주면 좋을 것 같아요.

kyukong replied:
피드백과는 다른 내용이지만 `검증 로직을 어디에 두어야 할까?` 가 이번 리팩토링에서 가장 고민했던 부분이었습니다. 

처음 PR을 보낼 때에는 `Input` 에 뒀었습니다. 입력값에 대한 검증이니까 입력 클래스와 같이 두는 것이 적절하다 생각했습니다. 그러나 Input에 두다보니 테스트 코드를 작성할 때 private 메소드에 접근해야 하는 상황이 벌어져 테스트 해야한다고 생각되는 메소드들을 public 으로 바꾸거나 테스트 코드로 옮겼습니다. 
이를 수정하고자 이번 리팩토링 시 `Validator` 객체를 만들어 검증 로직을 분리하였습니다. 별도의 객체로 분리하니 메인 메소드가 public 이 되어 테스트하기 더 쉬운 환경이 되었습니다. 하지만 모든 값에 대한 검증이 한 클래스 내에 존재하는 것이 가독성도 떨어졌습니다. 
페어와 각자의 피드백을 공유하면서 검증 로직에 대해 의견을 나누었는데 페어는 객체를 생성하는 클래스에 검증 로직을 두었다고 했습니다. 객체의 생성자는 public 이라서 다른 사용자가 무분별하게 인스턴스를 생성할 수 있는데 그 때 검증을 하고 검증된 값에 한해서 인스턴스를 생성해야 한다 했습니다. 들어보니 그것이 맞다고 생각되어 `생성자`에서 검증하는 메소드를 호출하는 방식으로 코드를 수정하였습니다.

그러자 카일이 남겨주신 피드백이 생각이 났습니다. Input 에서 입력에 대한 검증이 있었는데 따로 피드백을 해주신 이유가 뭘까 고민해봤고 객체를 생성할 때 검증을 하는 것이 적절하다는 의미로 작성해주신 것이 아닐까 했습니다. 제가 이해한 것이 맞는지는 모르겠지만 덕분에 검증 로직 위치에 대해 고민해보는 시간을 가질 수 있었습니다. 😀😀

KIMSIYOUNG replied:
생각의 흐름이 자연스럽네요🙂 좋은 고민인 것 같습니다.   
규콩님이 생각하신 것 처럼 `시도 횟수 1이상에 대한 검증`은 저도 도메인에 넣어주시길 생각하며 남긴 피드백이 맞습니다. 추가적으로 조금 더 코멘트 드리자면 앞으로 많은 작성하시는 프로그램에 `검증`에 대한 많은 요구사항이 존재할텐데, 이 때 이 검증 로직들은 `각각 어디에 위치하는게 맞을까` 라는 고민을 계속해서 이어가시면서 본인만의 기준을 갖는 것이 좋을 것 같아요. 각각이라고 표현한 것은 `검증`을 하나로 보지 않고 검증내에서 다시 분류가 될 수 있다는 의미이고, 어떤 검증들은 어디에 위치해야하고, 또 다른 어떤 검증들은 어디에 위치하는게 좋은 것 같다라는 기준들을 계속해서 구체화해나가시면 좋을 것 같습니다.

kyukong replied:
메소드가 가지는 성격별로 클래스를 나누는 것이 옳다고 생각했었습니다. 😅 피드백을 통해서 클래스를 나누는 것은 결국 유지보수, 접근을 제한하기 위해서 등 여러 이유가 있을 것이고 무조건 분리만 한다고 편리한 코드가 되지 않는다는 것을 알았습니다! 다음 미션들을 진행하면서 검증에 대한 것들이 더 추가될텐데 그때 카일님이 남겨주신 피드백에 대해 다시 고민해보고 코드를 작성하겠습니다. 감사합니다!! 😀

KIMSIYOUNG commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,48 @@
+package racingcar.domain;
+
+public class Car implements Comparable<Car> {
```
자동차 포지션의 최대값을 구하기 위해 `Comparable`을 사용해주신 것 같아요👏

이는 Getter 사용을 지양하기 위함으로 추측되는데, 왜 Getter를 지양해야하는지도 고민해보면 좋을 것 같아요. 

Comparable 활용 👏

kyukong replied:
getter 와 setter 를 지양해야하는 것은 객체의 private 변수에 직접 접근할 수 있기 때문이라고 생각합니다. 공개/수정되어서는 안되는 값에 누구나 접근을 할 수 있다면 해당 객체에 대한 신뢰가 떨어집니다.

되도록 지양하고는 싶었지만 최종 우승자를 출력하는 과정에서 Output 에게 우승자의 이름과 위치정보를 넘겨줄 수밖에 없는 코드를 짜서 조금 아쉽습니다...🥲

KIMSIYOUNG replied:
좋은 고민인 것 같아요. 그렇다면 Getter를 사용하는 것이 공개/수정되어선 안되는 값에 누구나 접근할 수 있다와 같은 말일까요? 

상태를 가지고 있는 객체가 자신의 상태를 표현하는 것은 매우 자연스럽지만, 도메인객체가 뷰와 관련된 로직을 가지는 것에 대한 단점도 분명 존재하는데요. 이 부분도 조금 더 이어서 고민해보시면 좋을 것 같아요. 프로그램의 핵심적인 기능을 담당하는 `자동차`라는 객체가 `뷰(사용자에게 보여지기 위한 로직)`을 담당하고 있다면, 문제가 되는 것이 무엇일까? 라고 질문해보며 답을 찾아가면 좋을 것 같아요.  

조금 다른 이야기로 Getter로 값을 외부에 공개하지만 그 값이 바뀌지 않도록 하는 방법도 있습니다. 아래의 글도 참고로 읽어보면 좋을 것 같아요.
https://tecoble.techcourse.co.kr/post/2020-05-18-immutable-object/

KIMSIYOUNG commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,48 @@
+		this.name = name;
+	}
+
+	private void move() {
```
취향차이지만 프라이빗 메소드와 퍼블릭 메소드의 위치에 대해서도 고민해보면 좋을 것 같아요. 

해당 로직을 읽고 유지보수 하는 개발자가 보다 빠르게 로직들을 파악해나가기 위해서, 어떤 위치에 두면 좋을까? 라는 관점에서 고민해보고 유콩의 생각을 공유해주면 좋을 것 같습니다.

kyukong replied:
페어와 말로 의견을 주고 받다보니 코드를 작성하는 위치를 신경쓰지 못했네요.😅 
외부에서 사용 가능한 public 을 가장 상단에 두고 그에 속해있는 private 메소드들을 호출되는 순서대로 작성하는 것이 개발자 입장에서도 사용자 입장에서도 편할거라 생각합니다. 관련 코드는 수정하였습니다!

KIMSIYOUNG commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	@Override
+	public String toString() {
```
일반적으로 View와 도메인의 로직은 분리하는 것이 좋다고 하는데요. 

그 이유에 대해서 고민해보고 공감한다면 수정해보는 것도 좋을 것 같아요. 도메인에서 뷰를 위한 로직이 있어도 된다고 생각하시면 그 이유도 같이 알려주시면 좋을 것 같아요🙂

kyukong replied:
toString() 메소드가 문자열을 Output 에게 "넘겨준다" 라고 생각해서 View 일것이라고 생각하지 못했습니다. 🥲 아직 도메인과 뷰의 개념이 헷갈리는듯합니다. 

일단 제가 생각하는 도메인의 개념은 실존하는 객체를 나타낸 것으로, 상태값을 변경하는 메소드만 존재하는 것이 옳다고 생각합니다. 

뷰는 말 그대로 해당 프로그램을 사용하는 사용자가 마주치는 메소드와 관련된 그룹입니다. 입력 또는 출력이 있으며 출력 시 도메인으로부터 상태값을 받아와 해당 정보를 사용자에게 보여줍니다. 제가 오버라이딩한 toString() 은 상태값이 아닌 상태값을 가공한 문자열을 받아와 그대로 출력하므로 뷰의 성격에 가까워 getter 로 받아온 값을 output 에서 가공하여 출력하는 것으로 수정하였습니다.

KIMSIYOUNG replied:
`View와 도메인이 개념적으로 혹은 구현과정에서 어떠어떠해야 한다` 라는 기준이 모호하시다고 해주신 것 같아요🙂 

그 기준이 조금 더 명확해질 때 자동차의 위치를 출력하는 로직은 어디에 있어야하는지도 명확해질 것 같아요. 2단계에서도 계속 이어서 고민해주시면 좋을 것 같습니다. 아래의 고민들로 이어나가시면 좋을 것 같습니다.
- View라는 이름이 있기 때문에 화면 출력 로직은 뷰에 있어야하는 것인가
- 그렇다면 도메인이 해당 상태를 관리하는 주체적 객체인데 왜 뷰의 로직을 가지면 안되는건가?
- 어떤 단점이 있기에 상태를 가지고 있는 객체가 자신의 상태를 표현하지 못하는 것일까
- 그 단점이 각자의 상태를 스스로 노출하고 관리하는 장점보다 큰걸까?

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+	private List<Car> cars;
+
+	public Cars(String[] names) {
+		cars = new ArrayList<>();
```
인스턴스 변수를 선언과 동시에 초기화하지 않고, 생성자에서 이를 초기화하는 것에는 어떤 차이가 있을까요?

유콩은 생성자를 통해 초기화하는 것을 택한 것 같은데 이유도 알려주시면 좋을 것 같아요!

kyukong replied:
창피하지만 페어와 같이 진행을 하면서 코드를 작성할 때 신경을 쓰지 못했습니다 🥲 만약 제가 코드를 썼다면 인스턴스 변수를 선언과 동시에 초기화를 했겠지만 그건 습관적으로 작성했을듯 합니다. 카일이 피드백으로 남겨주어 인스턴스 변수의 초기화 위치에 대해 고민해보았습니다. 

생성자 안에 초기화를 한다면 List 와 같은 타입의 값을 ArrayList 나 LinkedList 등 여러 타입의 값을 받을 수 있을듯합니다. 입력값에 따라 생성자의 내용을 달리하고 그 상황에 맞춰 타입을 다르게 받아야 한다면 생성자 안에서 초기화를 해 활용하면 좋을것같습니다.

KIMSIYOUNG replied:
인터페이스의 개념을 잘 이해하고 있으시네요👏 좋은 관점인 것 같습니다.

말씀주신대로 습관적으로 작성하는 코드들은 규콩에게는 눈에 익은 형태이겠지만 누군가에겐 아닐 수 있기때문에, 지금처럼 습관적인 코드들에 대해서도 의심해보고 왜 이렇게했지 라는 고민을 계속하면 좋을 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.stream.Collectors;
+
+public class Cars {
+	private static final String MAX_FIND_ERROR = "max 값을 찾을 수 없습니다.";
```
`해당 문자열이 매직넘버`인가? 라는 것에 대해 고민해보면 좋을 것 같아요.

어떤 값들을 상수화해야하는지에 대한 유콩만의 기준을 세워가는 과정이라고 생각해주시고 이 값이 이렇게 포장되어야 하는 이유가 무엇인지도 고민해보면 좋을 것 같아요.

kyukong replied:
카일의 피드백을 보고 `나는 왜 모든 문자열 및 숫자를 상수화했지?` 를 고민해봤습니다. 그저 저는 상수화의 기준이 `나중에 수정될 가능성이 있는가?` 에만 초점을 맞추었던것 같습니다. 카일의 피드백을 받고 상수화 기준에 대해 다시 고민해보았습니다. 😁

숫자 또는 문자열을 사용할 때 상수화하면 좋을것같은 기준을 정리해보았습니다.
1. 다시 사용할 가능성이 있는가?
여러번 사용되는 단어 및 숫자라면 상수화를 하는 것이 적절하다고 생각합니다. 예시로 자동차 이름은 5글자까지 허용하고 해당 정보를 에러메시지에도 띄워야하므로 숫자 5를 상수로 설정하였습니다.

2. 현재 프로그램에 의미가 있는 값인가?
제가 구현하려는 메소드가 음수임을 판단하는 메소드였다면 0보다 작은지 판단하는 코드에서 숫자 0은 상수화를 하지 않을 것입니다. 이미 해당 메소드는 `음수임을 판단하는 메소드` 라는 역할이 부여되어 있으므로 기준이 되는 0은 변하지 않는 값이기 때문입니다. 
자동차 경주 미션에서 경주하는 자동차의 이동 여부를 판단할 때 0에서 9까지의 값 중 랜덤으로 뽑은 값이 4이상이면 현재 자동차의 위치를 1 이동시킵니다. 여기서 0과 9와 3또는 4는 현재 프로그램에 의미가 있는 값이므로 상수화하였습니다. 

3. 여러 클래스에 걸쳐 사용되는가?
제가 PR 보낼 당시 남겼던 질문 중 하나가 클래스 내 상수로 사용하는 것이 좋은지, Enum 클래스를 사용하는 것이 좋은지, 였습니다. 위에 상수화에 대해 고민해보면서 `그럼 Enum 클래스는 어떤 경우에 사용하는 것이 좋을까?` 하는 생각이 들었고 특정 값이 여러 클래스 내에서 사용이 된다면 Enum 클래스를 고려해도 좋다고 생각이 되었습니다. 특정 상수에 대해서만 접근을 허용하는 것이 아닌 Enum 클래스로 생성하여 값을 호출하여 사용한다면 더 깔끔한 코드가 되지 않을까 합니다.

KIMSIYOUNG replied:
많은 고민을 해주셨네요👏 위의 이유들에 대해서 저도 공감하고 있습니다. 

다만 각 기준이 모두 성립해야 상수가 될 수 있다는 아니라는 점도 기억하고 넘어가시면 좋을 것 같습니다. 예를들어 한곳에서만 사용하지만 그 값이 매직넘버라 유의미한 이름을 가져야한다면 2번만 만족하지만 상수가 될 수 있는 것 처럼요! (이미 잘 이해하고 계신 것 같아요🙂)

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,60 @@
+		return new Car(name.trim());
+	}
+
+	public List<String> findWinners() {
```
자동차를 관리하는 `Cars`에서 Winner를 찾아오는 로직이 굉장히 자연스럽네요👏

KIMSIYOUNG commented on src/main/java/racingcar/service/RacingCarGame.java:
```diff
@@ -0,0 +1,45 @@
+	private Cars cars;
+	private Attempt attempt;
+
+	private final Input input;
```
자동차미션의 주제와 약간은 벗어날 수 있지만, 겹치는 부분도 있을 것 같아 코멘트 드려요!

`Input`이라는 변수를 내부 생성자에서 초기화하는 것과, 외부에서 해당 값을 파라미터로 주입해주는 것에는 어떤 차이가 있을까요? 랜덤을 테스트하는 것과 이어서 고민해보면 좋을 것 같아요

kyukong replied:
제가 질문드렸던 부분인데 아무리 생각해도 랜덤 테스트와의 연관성을 찾지 못했습니다...😂 2단계, 3단계 리팩토링을 진행하면서 다시 생각해보겠습니다....🥲🥲🥲

KIMSIYOUNG replied:
넵넵! 제가 말을 굉장히 추상적으로 남겼는데요. 2단계 미션하시면서 아마 조금 더 이해하실 수 있을 것 같습니다🙂 혹시 그렇지 않다면 말씀해주세요!

KIMSIYOUNG reviewed (changes requested):
안녕하세요 유콩🙂

첫 미션임에도 불구하고 메소드를 잘 분리해주셔서 코드를 파악하기가 쉬웠어요!
몇 가지 조금 더 고민해보면 좋을 것들이 있어 코멘트 남겼습니다. 
이해가 잘 되지 않거나, 논의하면 좋을 것들은 언제든 말씀주세요!

1. 프리코스 진행 당시 랜덤에 나올 값을 고정하여 추측하고 테스트 케이스를 실행하는 메소드가 있었는데,
혹시 다음과 같은 테스트 메소드를 만든다면 어떤 것을 찾아보는 것이 좋을지 여쭈어보고 싶습니다.
- 유콩과 수달의 아이디어로 좋은 방법을 생각하신 것 같아요. 다만 해당 로직을 테스트 하기위해서 프로덕션 코드를 테스트코드에 복사해서 테스트를 진행하신 것 같아요. 그렇게 할 수 밖에 없던 이유에 대해 고민해보고, Input과 관련된 코멘트를 남겼는데 그 부분과 같이 생각해본 뒤이야기 나눠보면 좋을 것 같습니다. `랜덤을 테스트 한다`라는 말의 의미에 대해서도 조금 더 고민해보면 좋을 것 같아요.

2. 상수의 개수가 증가함에 따라 해당 클래스 내에 선언하는 것이 좋은지, Enum 클래스로 분리하는 것이 좋은지, 아니면 또 다른 방법이 있는지 알고 싶습니다.
- 저 또한 유콩이 말씀주신 방법대로 상수를 사용하고 있습니다. 상수를 사용하는 다른 방법에 대해 고민하기보다, 유콩이 말씀하신 `한 클래스에 모아두는 것`과 각 클래스가 가지고 있는 것에 대한 차이와 장단을 인지하고 있으면 유콩의 기준대로 사용하시면 될 것 같아요🙂

kyukong committed:
docs: 프로그램 동작 기능 설명 추가

kyukong committed:
docs: 기능 요구사항 추가

kyukong committed:
refactor(Calculator): 메소드 네이밍 수정, int 배열로 변환하는 로직 수정

- splitAndSum() : validate, split, sum 의 단계로 나누어지도록 수정
- stringToIntArray() : stream 을 사용하여 코드 리팩토링
- sum() : 메소드의 역할을 명확하게 표시하도록 메소드 이름 수정

kyukong committed:
refactor: 상수화 축소

kyukong committed:
refactor: 검증 로직 클래스 분리

테스트가 필요한 메소드를 접근 제어자 public 으로 부여하기 위해 Input 에서 입력받은 값을 검증하던 로직을 별도의 Validator 클래스를 만들어 클래스를 분리함

kyukong committed:
refactor: 자동차 이름 검증 로직 클래스 분리

kyukong committed:
refactor: 검증을 Validator 에게 부여하여 기존 InputTest 를 ValidatorTest 로 수정

kyukong committed:
refactor: 시도횟수 검증 로직을 domain/Attempt 로 이동

- 객체 생성 시 올바른 값을 입력했는지 검증하기 위해 검증 로직을 해당 객체로 이동
- 그에 따른 테스트 코드도 domain/AttemptTest 로 이동

kyukong committed:
refactor: 자동차 검증 로직을 domain/Car 와 domain/Cars 로 이동

kyukong committed:
refactor(CarTest): Car의 drive 테스트 메소드 수정

kyukong committed:
refactor(Car): public 과 private 메소드 위치 조정

kyukong committed:
refactor(Cars): public 과 private 메소드 위치 조정

kyukong committed:
refactor(Cars): 인스턴스 변수 cars 초기화 위치 수정

kyukong committed:
fix(Cars): 생성자 내 생성자 호출 시 this 로 수정

생성된 인스턴스가 아닌 또 다른 생성자를 호출하여 제대로 값이 저장되지 않아 수정함

kyukong committed:
refactor: Car 객체의 toString 메소드 오버라이딩 삭제

KIMSIYOUNG commented on src/main/java/racingcar/domain/Attempt.java:
```diff
@@ -0,0 +1,45 @@
+        this.attempt = attempt;
+    }
+
+    public Attempt(String attempt) {
```
생성자 오버로딩을 사용하셨네요👏 일반적으로 생성자 오버로딩을 사용할 때는 `주 생성자`와 `부 생성자`를 개념적으로 분리하여 `부 생성자`는 `주 생성자`를 부르는 형태를 주로 사용하는데요. 현재 유콩이 작성해주신 방법과, 위에서 말씀드린 방법이 무엇이 다른지 고민해보면 좋을 것 같아요. 

부생성자가 주생성자를 호출한다는 의미는 아래와 같은 코드가 될 것 같아요.
```java
public Attempt(String attempt) {
        checkValid(attempt);
        this(Integer.parseInt(attempt));
        
}
```

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	private boolean hasNext() {
+		return generate() > DRIVE_FLAG;
```
지금의 형태도 너무 좋지만, 2단계에서는 요 부분을 어떻게 변경하면 좋을지 계속해서 고민해보면 좋을 것 같습니다🙂

자동차가 이동해야한다 아니다의 기준이 `Car`가 아닌 `Cars`에 있는게 맞을까?
- 아니라면 어떻게 `Car`에게 이 역할을 맡길 수 있을까?

KIMSIYOUNG commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -0,0 +1,84 @@
+	private List<Car> cars = new ArrayList<>();
+
+	public Cars(String names) {
+		this(names.split(DELIMITER));
```
여기서는 주생성자를 부르고 있네요🙂 장단에 대해서도 잘 아실 것 같네요👏

KIMSIYOUNG commented on src/main/java/racingcar/view/Output.java:
```diff
@@ -0,0 +1,44 @@
+	private static final String STEP = "-";
+	private static StringBuilder stringBuilder = new StringBuilder();
+
+	private Output() {
```
사소한 디테일에서 코드에 많은 고민을 하고 계신게 느껴지네요👏

KIMSIYOUNG commented on src/main/java/racingcar/view/Output.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Output {
+	private static final String STEP = "-";
+	private static StringBuilder stringBuilder = new StringBuilder();
```
스트링빌더를 사용하신 부분👏

다만 이렇게 선언되어 있는 경우 프로그램에서 1번을 부른 후 2번을 부른다면 어떻게 될지, 그리고 그 결과가 유콩이 원하는건지 확인해보면 좋을 것 같아요.

1번. 자동차의 이름을 빌더로 만들어 출력하는 메소드
2번. 자동차의 위치를 빌더로 만들어 출력하는 메소드

KIMSIYOUNG reviewed (approved):
안녕하세요 유콩. 
피드백을 잘 반영해주셨네요 🙂
너무 잘 구현해주셔서 1단계는 머지하면 좋을 것 같아요. 
조금 더 고민해보면 좋을 것들은 코멘트 남겼는데 2단계에서 같이 반영해주세요! 

고생하셨습니다.
